### PR TITLE
[SPARK-23062][SQL] Improve EXCEPT documentation

### PR DESCRIPTION
**Pull request by henryr**
## What changes were proposed in this pull request?

Make the default behavior of EXCEPT (i.e. EXCEPT DISTINCT) more
explicit in the documentation, and call out the change in behavior
from 1.x.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Could you update [R document](https://github.com/apache/spark/blob/master/R/pkg/R/DataFrame.R#L2856), too?

**Comment by henryr** [comment]
Done, thanks for pointing that out!

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -2873,6 +2873,7 @@ setMethod("intersect",
 #' @rdname except
 #' @export
 #' @note except since 1.4.0
+#' @note behaviour changed from \code{EXCEPT ALL} to \code{EXCEPT DISTINCT} in 2.0.
```
Comment (HyukjinKwon) [inline R/pkg/R/DataFrame.R]: Ur.. I think we have use `@note` for version specification in SparkR. Just adding

```
Note: blabla
```
should be fine like other places.

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -1364,7 +1364,9 @@ def subtract(self, other):
         """ Return a new :class:`DataFrame` containing rows in this frame
         but not in another frame.
 
-        This is equivalent to `EXCEPT` in SQL.
+        This is equivalent to `EXCEPT DISTINCT` in SQL.
+
+        (Note: Before Spark 2.0, the behavior was equivalent to `EXCEPT ALL` in SQL.)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: In PySpark, we can use `.. note:: `. This makes the doc pretty :).

File: R/pkg/R/DataFrame.R (hunk 1 above)
Comment (felixcheung) [reply R/pkg/R/DataFrame.R]: I don't mind it either way, but to note:
- r doc order and whitespace is significant, if you use `#' Note:` you must put it after L2856, if you put an extra `#'` ie. empty line that it becomes the `Details` section, which might be the right place; see http://spark.apache.org/docs/latest/api/R/awaitTermination.html

Comment (felixcheung) [reply R/pkg/R/DataFrame.R]: ie.
```
#' but not in another SparkDataFrame. This is equivalent to \code{EXCEPT DISTINCT} in SQL.
#'
#' Note: Before Spark 2.0.0, the behavior was equivalent to `EXCEPT ALL` in SQL.
#'
#' @param x a SparkDataFrame.
```

File: python/pyspark/sql/dataframe.py (hunk 2 above)
Comment (felixcheung) [reply python/pyspark/sql/dataframe.py]: nit: `2.0` to `2.0.0`

**Comment by dongjoon-hyun** [comment]
Also, cc @gatorsmile .

File: python/pyspark/sql/dataframe.py (hunk 2 above)
Comment (gatorsmile) [reply python/pyspark/sql/dataframe.py]: Actually, before 2.0, it is not equivalent to EXCEPT ALL. For details, see the PR: https://github.com/apache/spark/pull/12736

File: R/pkg/R/DataFrame.R (hunk 1 above)
Comment (gatorsmile) [reply R/pkg/R/DataFrame.R]: Nit: This is wrong.

**Comment by felixcheung** [comment]
I see, from reading that PR I think perhaps we should reference migration guide in sql programming guide instead of putting the whole description here.

**Comment by gatorsmile** [comment]
Yeah, we should document the behavior changes, but that was just a bug fix for 100% following the semantics of ANSI-SQL `EXCEPT DISTINCT`

**Comment by felixcheung** [comment]
@henryr could you update this PR to only include EXCEPT DISTINCT without the notes

**Comment by gatorsmile** [comment]
Just FYI, in ANSI SQL, `EXCEPT` = `EXCEPT DISTINCT`

**Comment by henryr** [comment]
Thanks all for the pointers and feedback! I've removed the references to the behavior before 2.0, and now the changes just make it explicit that this is `EXCEPT DISTINCT` (I appreciate that that's the meaning of `EXCEPT` per ANSI, but the behavior change since 1.x has confused users I've spoken to so seems worthwhile to make the documentation as clear as possible).

**Comment by gatorsmile** [comment]
@henryr Since Spark 2.3, Spark SQL documents all the behavior changes in [Migration Guides](https://spark.apache.org/docs/latest/sql-programming-guide.html#migration-guide). Hopefully, this can help our end users.

**Comment by gatorsmile** [comment]
LGTM

**Pull request closed without merging**